#
# Import Libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.linear_model import PassiveAggressiveClassifier

In [2]:
df = pd.read_csv('news.csv')
len(df)

6335

In [3]:
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
df.label.unique()

array(['FAKE', 'REAL'], dtype=object)

In [5]:
df.label.value_counts() # Balanced data

REAL    3171
FAKE    3164
Name: label, dtype: int64

# Split the dataset

In [6]:
X_train, X_test, y_train, y_test = train_test_split(df['text'],df['label'], test_size= 0.2, random_state=0)

# Start TF-IDF Vectorizer

In [7]:
# will set max_df to 0.7 to ignore terms that appear in more than 70% of the documents, and remove stop words

tfidf_vec = TfidfVectorizer(max_df= 0.7, stop_words='english')

In [8]:
tfidf_train = tfidf_vec.fit_transform(X_train)
tfidf_test = tfidf_vec.transform(X_test)

# Initialize  PassiveAggressiveClassifier

In [9]:
pac = PassiveAggressiveClassifier(max_iter=50)

In [10]:
pac.fit(tfidf_train,y_train)

PassiveAggressiveClassifier(max_iter=50)

In [11]:
y_pred = pac.predict(tfidf_test)

In [12]:
acc = accuracy_score(y_test, y_pred)
print(f"Accuracy Score is: {round(acc*100,2)}%")

Accuracy Score is: 93.45%


In [13]:
confusion_matrix(y_test, y_pred, labels=['FAKE', 'REAL'])

array([[570,  45],
       [ 38, 614]], dtype=int64)

In [14]:
print(classification_report(y_test, y_pred, labels=['FAKE', 'REAL']))

              precision    recall  f1-score   support

        FAKE       0.94      0.93      0.93       615
        REAL       0.93      0.94      0.94       652

    accuracy                           0.93      1267
   macro avg       0.93      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267



### The model is good with overall accuracy 93.4%, and we can depend on it as the data is balanced
### But for more conformation we used confusion matrixs and classification report,
### and the result if good for all labels

#
# Thank You :)
#
#